### PR TITLE
A new way to upload pictures to oss.

### PR DESCRIPTION
**Pull request by xinyu-io**
## Description
OSS is An object storage service provided by Alibaba.

This extension is a easy way for you to upload a picture to OSS,

and return the url.

Just copy a image type file or screen capture successfully,

then open Raycast and typing oss.

## Screencast
upload-success

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

### CODE REVIEW

File: extensions/oss/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Upload your pictures to OSS
+
+![upload-success screenshots](./images/upload-success.png)
```
Comment (pernielsentikaer) [inline extensions/oss/README.md:3]: Could you rename the folder `images` to `media` instead

File: extensions/oss/package.json [hunk 2]
```diff
@@ -0,0 +1,89 @@
+{
+  "$schema": "https://www.raycast.com/schemas/extension.json",
+  "name": "oss",
+  "title": "OSS",
+  "description": "Publish the picture in your clipboard to aliyun oss, and return url.",
```
Comment (pernielsentikaer) [inline extensions/oss/package.json:5]:
```suggestion
  "description": "Publish the picture in your clipboard to Aliyun OSS, and return the URL.",
```

File: extensions/oss/package.json [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  "icon": "command-icon.png",
+  "author": "yangxy",
+  "license": "MIT",
+  "commands": [
+    {
+      "name": "index",
+      "title": "Pic Upload",
```
Comment (pernielsentikaer) [inline extensions/oss/package.json:12]:
```suggestion
      "title": "Upload Picture",
      "subtitle": "Aliyun OSS",
```

File: extensions/oss/package.json [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      "description": "Upload the most top picture in your clipboard.",
```
Comment (pernielsentikaer) [inline extensions/oss/package.json:13]:
```suggestion
      "description": "Upload picture from the clipboard.",
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @yangxyo 👋 

Thanks for your contribution, I have some comments and feedback for you to address

- The extension needs a proper icon which is not the default example icon
- Since It's not possible to test for me, can you record a screencast of the flow

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by yangxyo** [comment]
@pernielsentikaer 
I have solved all the problems you raised.
Please...

**Comment by pernielsentikaer** [comment]
Thanks, @yangxyo 💪 

Can you create video showing the extension, so I have a better idea of what's going on? 😄

**Comment by yangxyo** [comment]
https://user-images.githubusercontent.com/32320149/156487976-94ec9ec1-cb5f-4c90-af60-792314daecd0.mov

File: extensions/oss/src/index.tsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    }
+    showToast({
+      style: Toast.Style.Failure,
+      title: title,
+      message: message || "Upload Image Failed",
+    });
+  }
+
+  return (
+    <List isLoading={!pic.url}>
```
Comment (pernielsentikaer) [inline extensions/oss/src/index.tsx:99]:
```suggestion
    <List isLoading={pic.url !== ""}>
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @yangxyo 👋 

Thanks for the video, it helped with the understanding 😄 

I have some feedback and comments for you to address:

- I would show a failure toast if clipboard is empty (lack image)
- Maybe rename the extension folder to something like`aliyun-oss`

We're getting closer, looking forward to testing it out again 😄 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by yangxyo** [comment]
image

Why do you would like to change `{!pic.url}` to `{pic.url !== ""}`?

**Comment by yangxyo** [comment]
I've changed oss to 阿里云-oss

**Comment by pernielsentikaer** [comment]
> Why would you like to change `{!pic.url}` to `{pic.url !== ""}`?

I'm not at my machine right now, but yeah, I'm totally wrong there 😄 

What you need to look into is when anything goes wrong, it never stops the loader:

https://user-images.githubusercontent.com/38033107/156766140-7b02d07a-9c79-4b33-8ebb-5c990fc14d4f.mp4

**Comment by yangxyo** [comment]
@pernielsentikaer 
Fixed.
image

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Much better @yangxyo 🔥 

Last feedback before merging 😄 

- Would be nice with a better toast when `pngpaste` is not installed, maybe check if path exists?
- No need to title case in toasts, could you write it like “Copy image failed”

When this is corrected, then I'm sure we're ready to ship 🚢

**Comment by yangxyo** [comment]
@pernielsentikaer 
🚀

File: extensions/aliyun-oss/package.json [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      "required": true,
+      "description": "Your pngpaste full path",
+      "default": "/opt/homebrew/bin/pngpaste"
+    },
+    {
+      "name": "domain",
+      "title": "Custom Domain Name",
+      "type": "textfield",
+      "required": false,
+      "placeholder": "http://raycast.com",
```
Comment (pernielsentikaer) [inline extensions/aliyun-oss/package.json:61]:
```suggestion
      "placeholder": "https://raycast.com",
```

File: extensions/aliyun-oss/src/index.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    const picName = getPicDataAndPicName();
+    if (picName) {
+      picUpload(picName).then((pic) => {
+        if (pic && pic.url) {
+          if (preferences.domain) {
+            try {
+              const url = new URL(pic.url);
+              pic.url = pic.url.replace(url.origin, preferences.domain);
+            } catch {
+              handleShowToast("Convert URL Failed.");
```
Comment (pernielsentikaer) [inline extensions/aliyun-oss/src/index.tsx:46]:
```suggestion
              handleShowToast("Convert URL failed.");
```

File: extensions/aliyun-oss/src/index.tsx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    if (fs.existsSync(preferences.pngpasteFullPath)) {
+      let picName = "";
+      try {
+        const { failed } = execaSync(preferences.pngpasteFullPath, ["/tmp/upload-to-oss"]);
+        if (!failed) {
+          const { stdout: name } = execaSync("pbpaste", [], { encoding: null });
+          picName = iconv.decode(Buffer.from(name), "cp936") || `${new Date().getTime()}.png`;
+        }
+      } catch {
+        handleShowToast("Copy Image Failed.");
```
Comment (pernielsentikaer) [inline extensions/aliyun-oss/src/index.tsx:65]:
```suggestion
        handleShowToast("Copy image failed.");
```

File: extensions/aliyun-oss/src/index.tsx [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+      }
+      return picName;
+    } else {
+      handleShowToast("The Path Of Pngpaste Is Wrong.");
```
Comment (pernielsentikaer) [inline extensions/aliyun-oss/src/index.tsx:69]:
```suggestion
      handleShowToast("The path ff pngpaste is wrong.");
```

File: extensions/aliyun-oss/src/index.tsx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+      setLoading(false);
+      handleShowToast("Upload Image Failed.");
+    }
+    return null;
+  }
+
+  function handleShowToast(title: string) {
+    showToast({
+      style: Toast.Style.Failure,
+      title: title,
```
Comment (pernielsentikaer) [inline extensions/aliyun-oss/src/index.tsx:97]:
```suggestion
      title: "An error occurred",
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @yangxyo 👋 

I have left some more feedback, we're getting pretty close now 💪 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by yangxyo** [comment]
@pernielsentikaer 
Ok.

**Review by pernielsentikaer — APPROVED** [review]
Hi @yangxyo 👋 

I think were ready to get it merged, and then we take eventually issues later on if anything arrives 💪 

Approved! 💯

**Comment by fe9lix** [comment]
Hey @yangxyo, the PR still contains the (old?) oss folder, in addition to the renamed folder. I guess that's a leftover and should be removed?

**Comment by yangxyo** [comment]
@fe9lix Sorry, I removed it.

**Pull request merged**